articles/055_ros_parameter_design.md

```diff
@@ -0,0 +1,200 @@
+
+### Unambigious naming
+
+All parameters should be unambiguously addressable.
+
+For example, one of the challenges of the current system is that there is a naming ambiguity between nodes and parameters `/foo/bar/baz` could be a node `/foo/bar/baz` or a private parameter `baz` on node `/foo/bar`.
+
+### Logging
+
+All changes to parameters should be visible when logging.
+And when playing back a log file the parameter changes should be able to be reapplied.
+
+## Proposed Approach
+
+To cover the feature set above, the ROS 2.0 parameter system is proposed as follows.
+
+### Parameters Hosted on nodes
+
+For the sake of validating parameter lifecycle, all parameters will be hosted on a node.
+Their lifetime will be implicitly tied to the nodes lifetime.
+The node will be responsible for validating current values.
+
+### Parameter addressing
+
+All parameters will be addressed by two elements: the full node name and the parameter name.
+
+### Supported datatypes
+
+Each parameter consists of a key and a value.
+The key is a string value.
+The value can be one of the following datatypes:
+
+- `float64`
+- `int64`
+- `string`
+- `bool`
+- `bytes[]`
+
+The datatypes are chosen as non-complex primatives.
+The full complement of datatypes of different bitdepth and unsigned types are avoided to allow interpretation from text based configuration files.
+
+`bytes` are included to allow the storage of binary blobs.
+It's use is not recommended but can be very convenient, and explicitly supporting it is better than having people try to abuse other datatypes such as strings.
+
+### Core API
+
+Each node is responsible for providing the following functionality.
+
+- **Get Parameters**
+  Given a list of parameter names it will return the parameter values.
+- **Set Parameters**
+  Given a list of parameter names, it will request an update of the values subject to validation of the values.
+  The updated values can include unsetting the value.
+  It will provide an API that can atomically update a set of values such that if any of the values fail validation, none of the values are set.
+  The success or failure of this call will be available to the client.
+- **List Parameters**
+  Provide a list of parameter names which are currently set.
+- **Describe Parameters**
+  Given a list of parameter names, return their datatype.
+
+This functionality will be exposed through local API calls as well as a ROS Service call API for remote operations.
+
+### Parameter update validation
+
+The node will allow for the registration of a callback for custom parameter value validation.
+
+### Backwards compatibility Parameter Server like behavior
+
+There are use cases where the older behavior with parameters persisting beyond the duration of a specific node are valuable.
+To this end we propose to write a simple node which emulates the policy of the ROS 1.0 parameter server: it runs in namespace `/` and simply accepts all change requests.
```

Comment:
I think this might be mixing two issues. One is the desire to have a global parameter space where no one owns the parameters, i.e. they are not validated by anyone. The other is to have parameters persist when the owner is shutdown and restarted, which can apply to either parameters owned and validated by a node or to a global parameter server.